### PR TITLE
Wr368432 remove libs cli

### PR DESCRIPTION
**Pull request by andrewmadden**
Adds a cli script to remove h5p libraries.

* There is no functionality in the UI for this.
* Can only remove libraries that have no other libraries depending on it.
* Can either use id or title and version to select the library
* Includes a dry run

### CODE REVIEW

File: classes/helper.php [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+ */
+class helper {
+
+    /**
+     * Get an installed library.
+     *
+     * @param array $options Array containing either id or a title and version number.
+     * @return \stdClass Record from hvp_libraries.
+     */
+    public static function get_library(array $options): \stdClass {
```
Comment (Peterburnett) [inline classes/helper.php:35]: Id like to see options become a little clearer. Can you start with a default options array, then foreach it and apply the supplied value if its present? Its very hard to track what makes a valid options array atm.

File: classes/helper.php [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        if (empty($dependentlibraries)) {
+            return $titles;
+        }
+
+        // If dependencies are found, get a list of titles with version number.
+        $dependencies = $DB->get_records('hvp_libraries_libraries', ['required_library_id' => $library->id],
+            '', 'library_id');
+        $dependencies = array_keys($dependencies);
+        list($insql, $params) = $DB->get_in_or_equal($dependencies);
+        $sql = "SELECT id, title, major_version, minor_version, patch_version
```
Comment (Peterburnett) [inline classes/helper.php:98]: Can we format this to be use the standard Moodle SQL formatting

File: cli/remove_library.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+// 1. Check if there is any dependency on the lib about to be deleted and stop process with message if one is found.
+$dependantlibraries = \mod_hvp\helper::get_dependant_libraries($library);
+if (!empty($dependantlibraries)) {
+    cli_error('The library can\'t be deleted as it is a dependency for other libraries: ' . implode(',', $dependantlibraries));
+} else {
+    cli_writeln("Library '{$library->title} {$library->major_version}.{$library->minor_version}.{$library->patch_version}' can be deleted.");
+}
+
+// 2. Search for all activities that are using the lib and delete them.
+try {
```
Comment (Peterburnett) [inline cli/remove_library.php:74]: This also needs to be gated behind the dryrun flag

Comment (Peterburnett) [reply cli/remove_library.php:74]: Ah I see that there is a dryrun passed through. There should still be an output here, currently the count is silent, so lets echo the count,

File: remove_library.php [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+require_once($CFG->dirroot . '/course/lib.php');
+require_once($CFG->dirroot . '/mod/hvp/lib.php');
+
+require_admin();
+
+$options = [
+    'help' => optional_param('help', false, PARAM_BOOL),
+    'library-id' => optional_param('library-id', '', PARAM_ALPHANUMEXT),
+    'library-title' => optional_param('library-title', '', PARAM_ALPHANUMEXT),
+    'major-version' => optional_param('major-version', '', PARAM_ALPHANUMEXT),
```
Comment (Peterburnett) [inline remove_library.php]: I think everywhere lets make these params a little more verbose and less painful to type, especially when doing bulk maintenance.

library-id -> id
library-title -> title
major-version -> major
minor-version -> minor
patch-version -> patch

File: remove_library.php [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    $dependantlibraries = \mod_hvp\helper::get_dependant_libraries($library);
+    if (!empty($dependantlibraries)) {
+        throw new moodle_exception('error:librarydependency', 'hvp', '', implode(',', $dependantlibraries));
+    } else {
+        $output .= html_writer::tag('p', "Library '{$library->title} {$library->major_version}.{$library->minor_version}.{$library->patch_version}' can be deleted.");
+    }
+
+    // 2. Search for all activities that are using the lib and delete them.
+    try {
+        $countactivities = \mod_hvp\helper::remove_dependant_activities($library, !$options['run']);
```
Comment (Peterburnett) [inline remove_library.php:64]: Same here, lets expose the activity count.

File: classes/helper.php [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+                'patch_version' => $options['patch-version'],
+            ]);
+        }
+
+        if (empty($libraries)) {
+            throw new \moodle_exception('error:nolibraryexists', 'hvp');
+        }
+
+        if (count($libraries) !== 1) {
+            throw new \moodle_exception('error:multiplelibrariesfound', 'hvp');
```
Comment (Peterburnett) [inline classes/helper.php:71]: I dont really think this should be an error condition. Infact, this is ideal for deleting any old versions etc

File: cli/remove_library.php [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+if ($options['help']) {
+    cli_writeln(get_string('cli:help', 'hvp'));
+    exit(0);
+}
+
+$framework = \mod_hvp\framework::instance('interface');
+
+// Get the installed library objects. TODO: Allow multiple ids.
+try {
+    $library = \mod_hvp\helper::get_library($options);
```
Comment (Peterburnett) [inline cli/remove_library.php:60]: If you always return an array from get_library (not allowing mutiple IDs, just allowing for blank versions to return all versions, then you can just wrap the whole rest of the file in a foreach loop

File: remove_library.php (hunk 4 above)
Comment (Peterburnett) [reply remove_library.php]: Also, the version input should be consistent with CLI. I prefer the --version=2.7.1, but that doesnt matter as much as long as the 2 methods are consistent

File: classes/helper.php (hunk 1 above)
Comment (andrewmadden) [reply classes/helper.php:35]: Good idea

File: classes/helper.php (hunk 2 above)
Comment (andrewmadden) [reply classes/helper.php:98]: Yep, good pick.

File: cli/remove_library.php (hunk 3 above)
Comment (andrewmadden) [reply cli/remove_library.php:74]: Message is here https://github.com/catalyst/moodle-mod_hvp/pull/20/files/85fd8a97890d2a881f6f020b9b78e82afffe10ae#diff-b214b8aa6a2164d2028c43a5f3a64e34cffe4f478cd99ea3c44ed7fc36388f06R80

File: remove_library.php (hunk 4 above)
Comment (andrewmadden) [reply remove_library.php]: Can't pass a period (`.`) as a query param in the url. I wanted to leave the --version=1.2.3 for the cli as it's much easier to use.

Will update the names, but leaving the versions as the full names as I don't think 'major' is descriptive enough.

File: remove_library.php (hunk 5 above)
Comment (andrewmadden) [reply remove_library.php:64]: Message here: https://github.com/catalyst/moodle-mod_hvp/pull/20/files/85fd8a97890d2a881f6f020b9b78e82afffe10ae#diff-49508de1afafd6780bd453b3dbfc9898b0fafc1ff6fbbd45dfb47b2ef53dc8e4R75

File: classes/helper.php (hunk 6 above)
Comment (andrewmadden) [reply classes/helper.php:71]: Due to how delicate the versions are with various dependencies, all library versions should not be deleted at the same time. Also, there is no uniqueness restriction with the title so technically 2 different libraries could have the same title and version.

File: cli/remove_library.php (hunk 7 above)
Comment (andrewmadden) [reply cli/remove_library.php:60]: As above, if doing bulk deletions, all libraries should be expressly given due to no uniqueness restrictions. For the amount of libraries to remove, I don't think this is a necessary feature to handle at this time.

File: classes/helper.php (hunk 1 above)
Comment (andrewmadden) [reply classes/helper.php:35]: Done

File: classes/helper.php (hunk 2 above)
Comment (andrewmadden) [reply classes/helper.php:98]: Done

File: remove_library.php (hunk 4 above)
Comment (andrewmadden) [reply remove_library.php]: Realised I needed to use PARAM_RAW, instead, for the optional params. Both scripts now only accept `id`, `title` and `version`.

File: classes/helper.php (hunk 6 above)
Comment (Peterburnett) [reply classes/helper.php:71]: No worries. I agree, if its potentially unsafe, lets just make it single.

File: cli/remove_library.php (hunk 3 above)
Comment (Peterburnett) [reply cli/remove_library.php:74]: Ah yep I somehow missed this

**Pull request merged**